src/rvsdg/simplify_branches.rs

```diff
@@ -0,0 +1,623 @@
+            let mut walker = self
+                .graph
+                .neighbors_directed(node, Direction::Outgoing)
+                .detach();
+            while let Some((edge, _)) = walker.next(&self.graph) {
+                if let BranchOp::Cond { arg, val, .. } = &self.graph[edge].op {
+                    res.branches
+                        .entry(arg.clone())
+                        .or_default()
+                        .insert(edge, *val);
+                }
+            }
+        }
+        res
+    }
+}
+
+#[derive(Default, Debug)]
+struct BranchMetadata {
+    /// Nodes that only contain administrative instructions.
+    admin_nodes: IndexSet<NodeIndex>,
+    /// Information about known constant values at particular nodes.
+    constants_known: ConstantInfo,
+    /// Relevant values used as branches.
+    branches: IndexMap<Identifier, IndexMap<EdgeIndex, CondVal>>,
+}
+
+/// Constants with a known value as of a given node.
+///
+/// For now, the constants are always booleans, but we keep arbitrary
+/// Literals around to make it easier to handle multi-way branches later.
+#[derive(Default, Debug)]
+struct ConstantInfo {
+    by_node: IndexMap<NodeIndex, IndexMap<Identifier, Literal>>,
+    by_id: IndexMap<Identifier, Vec<(NodeIndex, Literal)>>,
+}
+
+impl ConstantInfo {
+    fn add_constant(&mut self, node: NodeIndex, id: Identifier, lit: Literal) {
+        if self
+            .by_node
+            .entry(node)
+            .or_default()
+            .insert(id.clone(), lit.clone())
+            .is_none()
+        {
+            self.by_id.entry(id).or_default().push((node, lit));
+        }
+    }
+}
+
+/// "Administrative Instructions" are ones that will have essentially no runtime
+/// cost once they go through instruction selection / register allocation. We
+/// use these as a heuristic to find blocks that are safe to merge into their
+/// predecessors in exchange for simpler control flow: RVSDG conversion overhead
+/// is largely contained in blocks only containing these instructions.
+fn is_admin_instr(inst: &Instruction) -> bool {
+    matches!(
+        inst,
+        Instruction::Constant { .. }
+            | Instruction::Value {
+                op: ValueOps::Id,
+                ..
+            }
+    )
+}
+
+fn constants_assigned(inst: &Instruction) -> Option<(Identifier, Literal)> {
+    if let Instruction::Constant {
+        dest,
+        value: value @ Literal::Bool(..),
+        ..
+    } = inst
+    {
+        Some((dest.into(), value.clone()))
+    } else {
+        None
+    }
+}
+
+fn to_lit(cv: &CondVal) -> Option<Literal> {
+    if cv.of == 2 {
+        Some(if cv.val == 0 {
+            Literal::Bool(false)
+        } else {
+            Literal::Bool(true)
+        })
+    } else {
+        // Not handling multi-way branches for now.
+        None
+    }
+}
+
+/// A basic semilattice describing the state of a value.
+#[derive(Clone, Default, Debug)]
+enum ValueInfo {
+    #[default]
+    Bot,
```

Comment:
Worth commenting intuition about `Bot` and `Top`. I think `Bot` is "knows nothing" and `Top` is unreachable because we found a conflict